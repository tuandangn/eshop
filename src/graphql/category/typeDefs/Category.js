const { gql } = require('apollo-server');

const Category= gql`
    "Category Type"
    type Category implements ICategory & IPageableEntity{
        "ID"
        id: ID!,

        "Name"
        name: String!,

        "Type"
        type: CategoryType!,

        "Custom pagination"
        customPagination: Boolean!,

        "Default page size"
        defaultPageSize: Int!,

        "Page sizes"
        pageSizes: String,

        "Description"
        description: String,

        "Display order"
        displayOrder: Int!,

        "Show"
        show: Boolean!,

        "Parent"
        parent(
            "Allow hidden parent"
            showHidden: Boolean = false
        ): Category,

        "Children"
        children(
            "Include hidden children"
            showHidden: Boolean = false
        ): [Category!]!,
        
        "Products"
        products(
            "Include hidden products"
            showHidden: Boolean = false
        ): [IProduct!]!
    }
`;

module.exports = Category;