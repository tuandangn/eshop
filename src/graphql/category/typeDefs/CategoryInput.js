const { gql } = require('apollo-server');
const { DEFAULT_CATEGORY_CUSTOM_PAGINATION, DEFAULT_CATEGORY_PAGE_SIZE, DEFAULT_CATEGORY_SHOW, DEFAULT_CATEGORY_DISPLAY_ORDER } = require('../../../common/defaults');

const CategoryInput = gql`
    "Category Input Type"
    input CategoryInput{
        
        "Name"
        name: String!,

        "Type"
        type: CategoryType!,

        "Custom pagination"
        customPagination: Boolean = ${DEFAULT_CATEGORY_CUSTOM_PAGINATION},

        "Default page size"
        defaultPageSize: Int = ${DEFAULT_CATEGORY_PAGE_SIZE},

        "Page sizes"
        pageSizes: String,

        "Display order"
        displayOrder: Int = ${DEFAULT_CATEGORY_DISPLAY_ORDER},

        "Show"
        show: Boolean = ${DEFAULT_CATEGORY_SHOW},

        "Parent Category"
        parent: ID
    }
`;

module.exports = CategoryInput;