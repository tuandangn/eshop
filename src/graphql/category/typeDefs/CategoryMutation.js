const { gql } = require('apollo-server');

const Mutation = gql`
    "Category Mutation"
    type CategoryMutation {
        "Create"
        create(
            "Category"
            category: CategoryInput!
        ): Category!,

        "Delete"
        delete(
            "Category id"
            id: ID!
        ): Boolean!,

        "Update"
        update(
            "Category id"
            id: ID!,
            "Data"
            data: CategoryInput!
        ): Category!
    }
`;

module.exports = Mutation;