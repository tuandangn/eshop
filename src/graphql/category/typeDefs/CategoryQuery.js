const { gql } = require('apollo-server');

const CategoryQuery = gql`
    "Category Query"
    type CategoryQuery{
        "All"
        all(
            "Include hidden categories"
            showHidden: Boolean = false
        ): [Category!]!,

        "One"
        one(
            "Id"
            id: ID!,
            "Allow hidden category"
            showHidden: Boolean = false
        ): Category,

        "Search"
        search(
            "Name"
            name: String!,
            "Include hidden categories"
            showHidden: Boolean = false
        ): [Category!]!
    }
`;

module.exports = CategoryQuery;