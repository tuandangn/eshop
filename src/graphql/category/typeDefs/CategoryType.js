const { gql } = require('apollo-server');

const CategoryType = gql`
    "Category Type Enum"
    enum CategoryType{
        "Products"
        PRODUCTS,
        
        "News"
        NEWS
    }
`;

module.exports = CategoryType;