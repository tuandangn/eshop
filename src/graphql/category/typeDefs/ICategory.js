const { gql } = require('apollo-server');

const ICategory = gql`
    "Category Interface"
    interface ICategory {
        "Id"
        id: ID!,

        "Name"
        name: String!,

        "Type"
        type: CategoryType!,

        "Custom pagination"
        customPagination: Boolean!,

        "Default page size"
        defaultPageSize: Int!,

        "Page sizes"
        pageSizes: String,

        "Description"
        description: String,

        "Display order"
        displayOrder: Int!,
        
        "Show"
        show: Boolean!
    }
`;

module.exports = ICategory;