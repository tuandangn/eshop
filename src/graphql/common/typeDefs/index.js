module.exports = [
    require('./IPageInfo'),
    require('./IPageableEntity'),
    require('./IPagedList'),
    require('./PageInfo'),
    require('./SortBy'),
    require('./SortDirection'),
    require('./SortInput')
];