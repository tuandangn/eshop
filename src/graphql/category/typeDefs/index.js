module.exports = [
    require('./Category'),
    require('./CategoryType'),
    require('./ICategory'),
    require('./CategoryQuery'),
    require('./CategoryMutation'),
    require('./CategoryInput')
];