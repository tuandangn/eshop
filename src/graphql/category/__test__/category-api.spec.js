const { ApolloServer, gql } = require('apollo-server');
const { createTestClient } = require('apollo-server-testing');
const categoryResolver = require('../category.resolver');
const { NEWS, PRODUCTS } = require('../../../core/contants/category-type.const');
const { GROUP, SINGLE } = require('../../../core/contants/product-type.const');

const typeDefs = require('../typeDefs');
const RootQuery = gql`
    type Query{
        category: CategoryQuery
    }
`;
const dependencies = [
    require('../../product/typeDefs/IProduct'),
    require('../../product/typeDefs/Product'),
    require('../../product/typeDefs/ProductType'),
    require('../../common/typeDefs/IPageableEntity'),
    require('../../product/typeDefs/ProductCategoryType'),
];

const defaultResolver = {
    IProduct: {
        __resolveType: (obj) => "Product"

    },
    IPageableEntity: {
        __resolveType: (obj) => obj.price ? "Product" : "Category"
    }
}

const categoryFields = `
    id, name, type, customPagination, defaultPageSize,
    pageSizes, description, displayOrder, show
`;

describe('Category API', () => {
    describe('Category', () => {

        it('should query parent category', async () => {
            const allCategories = [
                { id: 1, name: 'category 1', type: NEWS, customPagination: false, defaultPageSize: 12, pageSizes: null, description: 'desc', displayOrder: 1, show: true, parent: undefined },
                { id: 2, name: 'category 2', type: PRODUCTS, customPagination: true, defaultPageSize: 10, pageSizes: '1,2,3,4', description: 'abcdef', displayOrder: 3, show: true, parent: 1 },
            ];
            const showHidden = false;
            const categoryService = {
                getAllCategories: jest.fn().mockResolvedValue(allCategories),
                getCategoryById: jest.fn().mockResolvedValue(allCategories[0])
            };
            const resolver = categoryResolver(categoryService);
            const server = new ApolloServer({
                typeDefs: [...typeDefs, ...dependencies, RootQuery],
                resolvers: [defaultResolver, resolver],
                mocks: true,
                mockEntireSchema: false
            });
            const { query } = createTestClient(server);

            const result = await query({
                query: gql`
                    query allCategories{
                        category{ 
                            all(showHidden: ${showHidden}){
                                parent{
                                    id, name
                                }
                            }
                        }
                }`
            });

            expect(result.errors).toBeFalsy();
            expect(result).toMatchSnapshot();
            expect(categoryService.getCategoryById).toHaveBeenCalledTimes(1);
            expect(categoryService.getCategoryById).toHaveBeenCalledWith(1, showHidden);
        });

        it('should query products by category', async () => {
            const category = {
                id: 1, name: 'category 1', type: NEWS,
                customPagination: false, defaultPageSize: 12, pageSizes: null,
                description: 'desc', displayOrder: 1, show: true, parent: undefined
            };
            const products = [{ id: "1", name: 'product 1', shortDesc: 'short', fullDesc: undefined, cost: undefined, oldPrice: undefined, type: SINGLE, individualVisibility: false, price: 0, displayOrder: 1, show: true }];
            const showHidden = false;
            const productShowHidden = true;
            const categoryService = {
                getAllCategories: jest.fn().mockResolvedValue([category]),
            };
            const productService = {
                getProductsInCategory: jest.fn().mockResolvedValue(products)
            };
            const resolver = categoryResolver(categoryService, productService);
            const server = new ApolloServer({
                typeDefs: [...typeDefs, ...dependencies, RootQuery],
                resolvers: [resolver, defaultResolver],
                mocks: true,
                mockEntireSchema: false
            });
            const { query } = createTestClient(server);

            const result = await query({
                query: gql`
                    query allCategories{
                        category{ 
                            all(showHidden: ${showHidden}){
                                id, name,
                                products(showHidden: ${productShowHidden}){
                                    id, name
                                }
                            }
                        }
                }`
            });

            expect(result.errors).toBeFalsy();
            expect(result).toMatchSnapshot();
            expect(productService.getProductsInCategory).toHaveBeenCalledWith(category.id, productShowHidden);
        });
    });

    describe('Query', () => {
        describe('all', () => {
            it('should query all categories with valid fields', async () => {
                const allCategories = [
                    { id: 1, name: 'category 1', type: NEWS, customPagination: false, defaultPageSize: 12, pageSizes: null, description: 'desc', displayOrder: 1, show: true, parent: undefined },
                    { id: 2, name: 'category 2', type: PRODUCTS, customPagination: true, defaultPageSize: 10, pageSizes: '1,2,3,4', description: 'abcdef', displayOrder: 3, show: false, parent: 1 },
                ];
                const showHidden = true;
                const categoryService = {
                    getAllCategories: jest.fn().mockResolvedValue(allCategories)
                };
                const resolver = categoryResolver(categoryService);
                const server = new ApolloServer({
                    typeDefs: [...typeDefs, ...dependencies, RootQuery],
                    resolvers: [resolver, defaultResolver],
                    mocks: true,
                    mockEntireSchema: false
                });
                const { query } = createTestClient(server);

                const result = await query({
                    query: gql`
                        query allCategories{
                            category{ 
                                all(showHidden: ${showHidden}){
                                    ${categoryFields}
                                }
                            }
                    }`
                });

                expect(result.errors).toBeFalsy();
                expect(result).toMatchSnapshot();
                expect(categoryService.getAllCategories).toHaveBeenCalledWith(showHidden);
            });
        });

        describe('one()', () => {
            it('should query one category', async () => {
                const category = {
                    id: 1, name: 'category 1', type: NEWS,
                    customPagination: false, defaultPageSize: 12, pageSizes: null,
                    description: 'desc', displayOrder: 1, show: true, parent: undefined
                };
                const id = "category id";
                const showHidden = false;
                const categoryService = {
                    getCategoryById: jest.fn().mockResolvedValue(category)
                };
                const resolver = categoryResolver(categoryService);
                const server = new ApolloServer({
                    typeDefs: [...typeDefs, ...dependencies, RootQuery],
                    resolvers: [defaultResolver, resolver],
                    mocks: true,
                    mockEntireSchema: false
                });
                const { query } = createTestClient(server);

                const result = await query({
                    query: gql`{
                        category{
                            one(id: "${id}", showHidden: ${showHidden}){
                                ${categoryFields}
                            }
                        }
                    }`
                });

                expect(result.errors).toBeFalsy();
                expect(result).toMatchSnapshot();
                expect(categoryService.getCategoryById).toHaveBeenCalledWith(id, showHidden);
            });
        });

        describe('search()', () => {
            it('should query search categories', async () => {
                const keywords = 'category';
                const showHidden = false;
                const categories = [{
                    id: 1, name: 'category 1', type: NEWS,
                    customPagination: false, defaultPageSize: 12, pageSizes: null,
                    description: 'desc', displayOrder: 1, show: true, parent: undefined
                }, {
                    id: 2, name: 'category 2', type: NEWS,
                    customPagination: false, defaultPageSize: 12, pageSizes: null,
                    description: 'desc', displayOrder: 1, show: true, parent: undefined
                }, {
                    id: 3, name: 'category 3', type: PRODUCTS,
                    customPagination: true, defaultPageSize: 9, pageSizes: '12, 9, 6, 4',
                    description: 'description', displayOrder: 5, show: true, parent: 1
                }];
                const categoryService = {
                    getCategoriesByName: jest.fn().mockResolvedValue(categories)
                };
                const resolver = categoryResolver(categoryService);
                const server = new ApolloServer({
                    typeDefs: [...typeDefs, ...dependencies, RootQuery],
                    resolvers: [defaultResolver, resolver],
                    mocks: true,
                    mockEntireSchema: false
                });
                const { query } = createTestClient(server);

                const result = await query({
                    query: gql`{
                        category{
                            search(name: "${keywords}", showHidden: ${showHidden}){
                                ${categoryFields}
                            }
                        }
                    }`
                });

                expect(result.errors).toBeFalsy();
                expect(result).toMatchSnapshot();
                expect(categoryService.getCategoriesByName).toHaveBeenCalledWith(keywords, showHidden);
            });
        });
    });
});