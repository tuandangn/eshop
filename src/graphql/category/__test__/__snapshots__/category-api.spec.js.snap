// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category API Category should query parent category 1`] = `
Object {
  "data": Object {
    "category": Object {
      "all": Array [
        Object {
          "parent": null,
        },
        Object {
          "parent": Object {
            "id": "1",
            "name": "category 1",
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Category API Category should query products by category 1`] = `
Object {
  "data": Object {
    "category": Object {
      "all": Array [
        Object {
          "id": "1",
          "name": "category 1",
          "products": Array [
            Object {
              "id": "1",
              "name": "product 1",
            },
          ],
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Category API Query all should query all categories with valid fields 1`] = `
Object {
  "data": Object {
    "category": Object {
      "all": Array [
        Object {
          "customPagination": false,
          "defaultPageSize": 12,
          "description": "desc",
          "displayOrder": 1,
          "id": "1",
          "name": "category 1",
          "pageSizes": null,
          "show": true,
          "type": "NEWS",
        },
        Object {
          "customPagination": true,
          "defaultPageSize": 10,
          "description": "abcdef",
          "displayOrder": 3,
          "id": "2",
          "name": "category 2",
          "pageSizes": "1,2,3,4",
          "show": false,
          "type": "PRODUCTS",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Category API Query one() should query one category 1`] = `
Object {
  "data": Object {
    "category": Object {
      "one": Object {
        "customPagination": false,
        "defaultPageSize": 12,
        "description": "desc",
        "displayOrder": 1,
        "id": "1",
        "name": "category 1",
        "pageSizes": null,
        "show": true,
        "type": "NEWS",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Category API Query search() should query search categories 1`] = `
Object {
  "data": Object {
    "category": Object {
      "search": Array [
        Object {
          "customPagination": false,
          "defaultPageSize": 12,
          "description": "desc",
          "displayOrder": 1,
          "id": "1",
          "name": "category 1",
          "pageSizes": null,
          "show": true,
          "type": "NEWS",
        },
        Object {
          "customPagination": false,
          "defaultPageSize": 12,
          "description": "desc",
          "displayOrder": 1,
          "id": "2",
          "name": "category 2",
          "pageSizes": null,
          "show": true,
          "type": "NEWS",
        },
        Object {
          "customPagination": true,
          "defaultPageSize": 9,
          "description": "description",
          "displayOrder": 5,
          "id": "3",
          "name": "category 3",
          "pageSizes": "12, 9, 6, 4",
          "show": true,
          "type": "PRODUCTS",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
